### PR TITLE
added support for warnings: fixes #1779

### LINKED ISSUE #1779: provide support for compiler warnings
required for #1366

### PR DESCRIPTION
**Pull request by sanderr**

### CODE REVIEW

**Comment by sanderr** [comment]
The classes `WarningRule`, `WarningOption` and `WarningsManager` in `inmanta.warnings` are an artefact of an ambitious start. They allow for pretty fine grained control of how warnings should be handled (warn, ignore, error). They basically add a translation layer between options specified as strings or booleans to Python's builtin warning rules. They are definitely not required for the current simple toggle on/off/error but they might be useful in the future when more fine grained warning control is desired.
@wouterdb, should I simplify this module by only handling the on/off/error options, or should it remain as is to provide a broader backend for future warning management?

File: src/inmanta/warnings.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+import warnings
+from typing import Dict, List, Mapping, Optional, Type, Union
+
+
+class WarningRule:
```
Comment (wouterdb) [inline src/inmanta/warnings.py]: This could use a bit of documentation. I have no idea what it does or why.

Comment (sanderr) [reply src/inmanta/warnings.py]: I'll add some documentation if you decide it should stay in (see my comment above). For now, it's a representation of Python's builtin warnings filter tuple: https://docs.python.org/3.8/library/warnings.html#the-warnings-filter. It indicates what action (ignore, warn, error, ...) should be taken for a warning that matches the rule's message, warning subclass, module and lineno where it occurred.

Comment (wouterdb) [reply src/inmanta/warnings.py]: I missed the comment. 

If the filters are used, they can stay, but I don't see where they are applied.

If they are dead code, they should be removed.

Comment (sanderr) [reply src/inmanta/warnings.py]: They are used, but only for a single option, where the rule is just instantiated as `WarningRule(desired_behaviour)` where `desired_behaviour` is one of `default`, `error`, `ignore`. This rule is a catch-all that toggles the desired behaviour for all warnings.
So I wouldn't really call it dead code, but I would call it overengineered for the currently limited application.

Comment (wouterdb) [reply src/inmanta/warnings.py]: Perhaps best to keep it simple, but no hard preference.

Comment (sanderr) [reply src/inmanta/warnings.py]: I kept the core of it to keep some abstraction but I simplified and added some documentation.

File: src/inmanta/warnings.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            {
+                "warn": WarningRule(WarningBehaviour.WARN),
+                "ignore": WarningRule(WarningBehaviour.IGNORE),
+                "error": WarningRule(WarningBehaviour.ERROR),
+            },
+        ),
+    ]
+
+    @classmethod
+    def apply_config(cls, config: Optional[Mapping]) -> None:
```
Comment (wouterdb) [inline src/inmanta/warnings.py]: could you type the mapping?

File: src/inmanta/warnings.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    def apply(self) -> None:
+        if self.type is None:
+            warnings.filterwarnings(self.action.value)
+        else:
+            warnings.filterwarnings(self.action.value, category=self.type)
+
+
+class WarningOption:
+    """
+        An option to manage warnings. Consists of a name and a range of possible values, each tied to a warning rule.
```
Comment (wouterdb) [inline src/inmanta/warnings.py]: Could you add some examples on how this could be used?

**Pull request merged**